unlock.go

```diff
@@ -115,52 +131,147 @@ type ReferentialUnlock interface {
+				if !unlock.Chainable() {
+					return ierrors.Wrapf(ErrReferentialUnlockInvalid, "%d references existing referential unlock %d but it does not support chaining", index, unlock.Ref())
 				}
-				seenRefUnlocks[uint16(index)] = x
+				seenRefUnlocks[uint16(index)] = unlock
 
 				break
 			}
-			// must reference a sig unlock here
-			if _, has := seenSigUnlocks[x.Ref()]; !has {
-				return ierrors.Wrapf(ErrReferentialUnlockInvalid, "%d references non existent unlock %d", index, x.Ref())
+
+			// must reference a sig or multi unlock here
+			_, hasSigUnlock := seenSigUnlocks[unlock.Ref()]
+			_, hasMultiUnlock := seenMultiUnlocks[unlock.Ref()]
+			if !hasSigUnlock && !hasMultiUnlock {
+				return ierrors.Wrapf(ErrReferentialUnlockInvalid, "%d references non existent unlock %d", index, unlock.Ref())
+			}
+			seenRefUnlocks[uint16(index)] = unlock
+
+		case *MultiUnlock:
+			multiUnlockBytes, err := api.Encode(unlock)
+			if err != nil {
+				return ierrors.Errorf("unable to serialize multi unlock block at index %d for dup check: %w", index, err)
+			}
+
+			if existingIndex, exists := seenMultiUnlockBytes[string(multiUnlockBytes)]; exists {
+				return ierrors.Wrapf(ErrMultiUnlockNotUnique, "multi unlock block at index %d is the same as %d", index, existingIndex)
+			}
+
+			for subIndex, subU := range unlock.Unlocks {
+				switch subUnlock := subU.(type) {
+				case *SignatureUnlock:
+					if subUnlock.Signature == nil {
+						return ierrors.Wrapf(ErrSigUnlockHasNilSig, "at index %d.%d is nil", index, subIndex)
+					}
+
+					sigBlockPubKeyBytes, err := publicKeyBytesFromSignatureBlock(subUnlock.Signature)
+					if err != nil {
+						return ierrors.Wrapf(err, "unable to parse pubkey bytes from signature unlock block at index %d.%d for dup check", index, subIndex)
+					}
+
+					// we check for duplicated pubkeys in SignatureUnlock(s)
+					if existingIndex, exists := seenSigBlockPubkeyBytes[string(sigBlockPubKeyBytes)]; exists {
+						return ierrors.Wrapf(ErrSigUnlockNotUnique, "signature unlock block at index %d.%d is the same as %d", index, subIndex, existingIndex)
+					}
+
+					// we don't set the index here in "seenSigUnlocks" because there is no concept of reference unlocks inside of multi unlocks
+
+					// add the pubkey to "seenSigBlockPubkeyBytesInMultiUnlocks", so we can check that pubkeys from a multi unlock are not reused in a normal SignatureUnlock
+					seenSigBlockPubkeyBytesInMultiUnlocks[string(sigBlockPubKeyBytes)] = index
+
+				case ReferentialUnlock:
+					if prevRef := seenRefUnlocks[subUnlock.Ref()]; prevRef != nil {
+						if !subUnlock.Chainable() {
+							return ierrors.Wrapf(ErrReferentialUnlockInvalid, "%d.%d references existing referential unlock %d but it does not support chaining", index, subIndex, subUnlock.Ref())
+						}
+						// we don't set the index here in "seenRefUnlocks" because there is no concept of reference unlocks inside of multi unlocks
```

Comment:
```suggestion
						// we don't set the index here in "seenRefUnlocks" because it's not possible to reference an unlock within a multi unlock
```

Would this be clearer perhaps?